File: rtl/tb/alu_conttb.cpp
#include <stdlib.h>
#include <iostream>
#include <cassert>
#include "Valu_cont.h"
#include <verilated.h>
#include <verilated_vcd_c.h>
#include <stdint.h>

#include "parameters.h"

#define MAX_CYCLES 6
vluint64_t sim_time = 0;

#define CHECK(signal, expected) \
    if ((signal) != (expected)) { \
        std::cerr << "ASSERT FAIL: " #signal " == " << (int)(signal) \
                  << ", expected " << (int)(expected) << std::endl; \
        exit(1); \
    }

void step(Valu_cont* alu_cont, VerilatedVcdC* trace, int cycles = MAX_CYCLES) {
    for (int i = 0; i < cycles; ++i) {
        alu_cont->eval();
        if (trace) trace->dump(sim_time);
        sim_time++;
    }
}

void test_rtype(Valu_cont* alu_cont, VerilatedVcdC* trace) {
    std::cout << "Requesting R-Type ALU ADD signal:\n";

    alu_cont->aluop = ALUOP_RTYPE;
    alu_cont->funct7 = 0;
    alu_cont->funct3 = 0;

    step(alu_cont, trace, 2);
    CHECK(alu_cont->alucontrol, ALU_ADD);


    std::cout << "Requesting R-Type ALU SUB signal:\n";

    alu_cont->aluop = ALUOP_RTYPE;
    alu_cont->funct7 = 32;
    alu_cont->funct3 = 0;

    step(alu_cont, trace, 2);
    CHECK(alu_cont->alucontrol, ALU_SUB);


    std::cout << "Requesting R-Type ALU AND signal:\n";

    alu_cont->aluop = ALUOP_RTYPE;
    alu_cont->funct7 = 0;
    alu_cont->funct3 = 7;

    step(alu_cont, trace, 2);
    CHECK(alu_cont->alucontrol, ALU_AND);


    std::cout << "Requesting R-Type ALU OR signal:\n";

    alu_cont->aluop = ALUOP_RTYPE;
    alu_cont->funct7 = 0;
    alu_cont->funct3 = 6;

    step(alu_cont, trace, 2);
    CHECK(alu_cont->alucontrol, ALU_OR);
}

void test_branch(Valu_cont* alu_cont, VerilatedVcdC* trace) {
    std::cout << "Requesting Branch ALU ADD signal:\n";

    alu_cont->aluop = ALUOP_BRANCH;

    step(alu_cont, trace, 2);
    CHECK(alu_cont->alucontrol, ALU_SUB);
}

void test_lwsw(Valu_cont* alu_cont, VerilatedVcdC* trace) {
    std::cout << "Requesting LWSW ALU ADD signal:\n";

    alu_cont->aluop = ALUOP_LWSW;

    step(alu_cont, trace, 2);
    CHECK(alu_cont->alucontrol, ALU_ADD);
}


int main(int argc, char **argv) {
    Verilated::commandArgs(argc, argv);
    Valu_cont* alu_cont = new Valu_cont;

    Verilated::traceEverOn(true);  // Enable VCD tracing
    VerilatedVcdC* m_trace = new VerilatedVcdC;
    alu_cont->trace(m_trace, 5);
    m_trace->open("waveform.vcd");

    test_rtype(alu_cont, m_trace);
    test_branch(alu_cont, m_trace);
    test_lwsw(alu_cont, m_trace);


    std::cout << "All tests passed.\n";

    m_trace->close();
    delete alu_cont;
    delete m_trace;
    return 0;
}

File: rtl/tb/alu_conttb.py
import cocotb
from cocotb.triggers import Timer, ReadOnly

from parameters import *

ALUOP_RTYPE  = PARAMS["ALUOP"]["ALUOP_RTYPE"]
ALUOP_BRANCH = PARAMS["ALUOP"]["ALUOP_BRANCH"]
ALUOP_LWSW   = PARAMS["ALUOP"]["ALUOP_LWSW"]

ALU_ADD = PARAMS["ALU_OPS"]["ALU_ADD"]
ALU_AND = PARAMS["ALU_OPS"]["ALU_AND"]
ALU_SUB = PARAMS["ALU_OPS"]["ALU_SUB"]
ALU_XOR = PARAMS["ALU_OPS"]["ALU_XOR"]
ALU_OR  = PARAMS["ALU_OPS"]["ALU_OR"]
ALU_SLT = PARAMS["ALU_OPS"]["ALU_SLT"]
ALU_SLTU = PARAMS["ALU_OPS"]["ALU_SLTU"]
ALU_SLL = PARAMS["ALU_OPS"]["ALU_SLL"]
ALU_SRL = PARAMS["ALU_OPS"]["ALU_SRL"]
ALU_SRA = PARAMS["ALU_OPS"]["ALU_SRA"]

async def request_type(dut, op, funct7, funct3, alucontrol, name=""):
    dut.aluop.value = op
    dut.funct7.value = funct7
    dut.funct3.value = funct3

    await ReadOnly()

    actual = dut.alucontrol.value
    assert actual == alucontrol, (
        f"{name} failed: "
        f"aluop={op:#x}, funct7={funct7:#x}, funct3={funct3:#x}, "
        f"expected alucontrol={alucontrol:#x}, got={alucontrol:#x}"
    )
    
    await Timer(10, "ps")


@cocotb.test()
async def test_rtype(alu_cont):
    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0,
        0,
        ALU_ADD,
        name="R-type ADD"
    )

    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0100000,
        0b0,
        ALU_SUB,
        name="R-type SUB"
    )

    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0,
        0b111,
        ALU_AND,
        name="R-type AND"
    )

    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0,
        0b110,
        ALU_OR,
        name="R-type OR"
    )

    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0,
        0b100,
        ALU_XOR,
        name="R-type XOR"
    )
    
    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0,
        0b001,
        ALU_SLL,
        name="R-type SLL"
    )

    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0,
        0b010,
        ALU_SLT,
        name="R-type SLT"
    )

    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0,
        0b011,
        ALU_SLTU,
        name="R-type SLTU"
    )

    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0100000,
        0b101,
        ALU_SRA,
        name="R-type SRA"
    )

    await request_type(
        alu_cont,
        ALUOP_RTYPE,
        0b0,
        0b101,
        ALU_SRL,
        name="R-type SRL"
    )



@cocotb.test()
async def test_lwsw(alu_cont):
    await request_type(
        alu_cont,
        ALUOP_LWSW,
        0b0,
        0b0,
        ALU_ADD
    )


@cocotb.test()
async def test_branch(alu_cont):
    await request_type(
        alu_cont,
        ALUOP_BRANCH,
        0b0,
        0b0,
        ALU_SUB
    )